From 0a37fcbbb55246459d01fd4e4c8336c76bbc872c Mon Sep 17 00:00:00 2001
From: Heinrich Schuchardt <xypron.glpk@gmx.de>
Date: Wed, 17 Jun 2020 14:41:45 +0200
Subject: [PATCH] efi_loader: simplify efi_set_variable_common()

Reduce the code size by simplifying efi_set_variable_common().
The size of u-boot.bin is reduced by 138 bytes for qemu_arm64_defconfig.

The internal representation of UEFI variables will have a trailing comma if
the variable is not authenticated.

Signed-off-by: Heinrich Schuchardt <xypron.glpk@gmx.de>
---
 lib/efi_loader/efi_variable.c | 36 +++++++++++------------------------
 1 file changed, 11 insertions(+), 25 deletions(-)

diff --git a/lib/efi_loader/efi_variable.c b/lib/efi_loader/efi_variable.c
index e097670e28..693529ff83 100644
--- a/lib/efi_loader/efi_variable.c
+++ b/lib/efi_loader/efi_variable.c
@@ -1011,32 +1011,18 @@ static efi_status_t efi_set_variable_common(u16 *variable_name,
 	/*
 	 * store attributes
 	 */
-	attributes &= (READ_ONLY |
-		       EFI_VARIABLE_NON_VOLATILE |
-		       EFI_VARIABLE_BOOTSERVICE_ACCESS |
-		       EFI_VARIABLE_RUNTIME_ACCESS |
-		       EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS);
 	s += sprintf(s, "{");
-	while (attributes) {
-		attr = 1 << (ffs(attributes) - 1);
-
-		if (attr == READ_ONLY) {
-			s += sprintf(s, "ro");
-		} else if (attr == EFI_VARIABLE_NON_VOLATILE) {
-			s += sprintf(s, "nv");
-		} else if (attr == EFI_VARIABLE_BOOTSERVICE_ACCESS) {
-			s += sprintf(s, "boot");
-		} else if (attr == EFI_VARIABLE_RUNTIME_ACCESS) {
-			s += sprintf(s, "run");
-		} else if (attr ==
-			   EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS) {
-			s += sprintf(s, "time=");
-			s = bin2hex(s, (u8 *)&time, sizeof(time));
-		}
-
-		attributes &= ~attr;
-		if (attributes)
-			s += sprintf(s, ",");
+	if (attributes & READ_ONLY)
+		s += sprintf(s, "ro,");
+	if (attributes & EFI_VARIABLE_NON_VOLATILE)
+		s += sprintf(s, "nv,");
+	if (attributes & EFI_VARIABLE_BOOTSERVICE_ACCESS)
+			s += sprintf(s, "boot,");
+	if (attributes &EFI_VARIABLE_RUNTIME_ACCESS)
+		s += sprintf(s, "run,");
+	if (attributes & EFI_VARIABLE_TIME_BASED_AUTHENTICATED_WRITE_ACCESS) {
+		s += sprintf(s, "time=");
+		s = bin2hex(s, (u8 *)&time, sizeof(time));
 	}
 	s += sprintf(s, "}");
 	s += sprintf(s, "(blob)");
-- 
2.27.0

